test/server/workers/testworkers.cpp

```diff
@@ -0,0 +1,317 @@
+    vfsSetupParams._userId = user.userId();
+#endif
+    vfsSetupParams._localPath = _sync.localPath();
+    vfsSetupParams._targetPath = _sync.targetPath();
+    vfsSetupParams._logger = _logger;
+
+#if defined(__APPLE__)
+    _vfsPtr = std::unique_ptr<VfsMac>(new VfsMac(vfsSetupParams));
+#elif defined(__WIN32)
+    _vfsPtr = std::unique_ptr<VfsWin>(new VfsWin(vfsSetupParams));
+#else
+    _vfsPtr = std::unique_ptr<VfsOff>(new VfsOff(vfsSetupParams));
+#endif
+
+    // Setup SyncPal
+    _syncPal = std::make_shared<SyncPal>(_sync.dbId(), KDRIVE_VERSION_STRING);
+    _syncPal->createWorkers();
+    _syncPal->syncDb()->setAutoDelete(true);
+    _syncPal->createProgressInfo();
+    _syncPal->setVfsCreatePlaceholderCallback(createPlaceholder);
+    _syncPal->setVfsConvertToPlaceholderCallback(convertToPlaceholder);
+    _syncPal->setVfsSetPinStateCallback(setPinState);
+}
+
+void TestWorkers::tearDown() {
+    ParmsDb::instance()->close();
+    ParmsDb::reset();
+    if (_syncPal && _syncPal->syncDb()) {
+        _syncPal->syncDb()->close();
+    }
+    if (_vfsPtr) {
+        _vfsPtr = nullptr;
+    }
+}
+
+void TestWorkers::testCreatePlaceholder() {
+    ExitCause exitCause{ExitCause::Unknown};
+    ExitCode exitCode{ExitCode::Unknown};
+
+    _syncPal->resetEstimateUpdates();
+
+    // Progress not intialized
+    {
+        exitCause = ExitCause::Unknown;
+        exitCode = _syncPal->_executorWorker->createPlaceholder(SyncPath("dummy"), exitCause);
+        CPPUNIT_ASSERT_EQUAL(exitCode, ExitCode::DataError);
+        CPPUNIT_ASSERT_EQUAL(exitCause, ExitCause::InvalidSnapshot);
+    }
+
+    // Create folder operation
+    SyncPath relativeFolderPath{"placeholders_folder"};
+    {
+        SyncFileItem syncItem;
+        syncItem.setPath(relativeFolderPath);
+        syncItem.setType(NodeType::Directory);
+        syncItem.setDirection(SyncDirection::Down);
+        _syncPal->initProgress(syncItem);
+
+        // Folder doesn't exist (normal case)
+        exitCause = ExitCause::Unknown;
+        exitCode = _syncPal->_executorWorker->createPlaceholder(relativeFolderPath, exitCause);
+        CPPUNIT_ASSERT_EQUAL(exitCode, ExitCode::Ok);
+        CPPUNIT_ASSERT_EQUAL(exitCause, ExitCause::Unknown);
+
+#if defined(__APPLE__) && defined(__WIN32)
+        // Folder already exists
+        exitCause = ExitCause::Unknown;
+        exitCode = _syncPal->_executorWorker->createPlaceholder(relativeFolderPath, exitCause);
+        CPPUNIT_ASSERT_EQUAL(exitCode, ExitCode::DataError);
+        CPPUNIT_ASSERT_EQUAL(exitCause, ExitCause::InvalidSnapshot);
+#endif
+    }
+
+    // Create file operation
+    SyncPath relativeFilePath{relativeFolderPath / "placeholder_file"};
+    {
+        SyncFileItem syncItem;
+        syncItem.setPath(relativeFilePath);
+        syncItem.setType(NodeType::File);
+        syncItem.setDirection(SyncDirection::Down);
+        _syncPal->initProgress(syncItem);
+
+#if defined(__APPLE__) && defined(__WIN32)
```

Comment:
This can never be true. Didn't you meant to use || oprerator?

Reply:
Oups. Fixed